### PR TITLE
Changes to (_internal_)create_program()

### PR DESCRIPTION
**Pull request by lumip**
Implementations of new signature PT._internal_create_program for RepetitionPT, SequencePT, ForLoopPT, AtomicPT (and thus all subclasses of that) + tests.

MappingPT is the only one missing now.
After that is done, #287 should be finished.

### CODE REVIEW

File: qctoolkit/pulses/pulse_template.py [hunk 1]
```diff
@@ -118,7 +118,7 @@ def create_program(self, *,
         if parameters is None:
             parameters = dict()
         if measurement_mapping is None:
-            measurement_mapping = dict()
+            measurement_mapping = dict() # todo (2018-08-02): should not be empty but an identity mapping; requires PT to be a MeasurementDefiner (?)
```
Comment (terrorfisch) [inline qctoolkit/pulses/pulse_template.py]: `measurement_names` is an abstract property of `PulseTemplate`

File: qctoolkit/pulses/pulse_template.py [hunk 2]
```diff
@@ -190,19 +197,25 @@ def build_sequence(self,
...
-                      for parameter_name, parameter_value in parameters.items()
-                      if parameter_name in self.parameter_names}
+        try:
+            parameters = {parameter_name: parameters[parameter_name].get_value()
+                          for parameter_name in self.parameter_names}
+
+            measurement_parameters = {parameter_name: parameters[parameter_name].get_value()
+                                      for parameter_name in self.measurement_parameters}
+        except KeyError as e:
+            raise ParameterNotProvidedException(str(e)) from e
```
Comment (terrorfisch) [inline qctoolkit/pulses/pulse_template.py]: Why do you create an exta dictionary for the measurement parameters? They should be a subset of `parameter_names`

File: qctoolkit/pulses/loop_pulse_template.py [hunk 3]
```diff
@@ -228,28 +228,30 @@ def build_sequence(self,
...
-                                  for parameter_name in self.measurement_parameters}
+        try:
+            measurement_parameters = {parameter_name: parameters[parameter_name].get_value()
+                                      for parameter_name in self.measurement_parameters}
+            duration_parameters = {parameter_name: parameters[parameter_name].get_value()
+                                      for parameter_name in self.duration.variables}
+        except KeyError as e:
+            raise ParameterNotProvidedException(str(e)) from e
         measurements = self.get_measurement_windows(measurement_parameters, measurement_mapping)
+        if self.duration.evaluate_numeric(**duration_parameters) > 0:
```
Comment (terrorfisch) [inline qctoolkit/pulses/loop_pulse_template.py]: There should be a better way. But this is an optimization for the future.

File: qctoolkit/pulses/pulse_template.py (hunk 2 above)
Comment (lumip) [reply qctoolkit/pulses/pulse_template.py]: should be but currently aren't according to https://github.com/qutech/qc-toolkit/issues/230

File: qctoolkit/pulses/pulse_template.py (hunk 1 above)
Comment (lumip) [reply qctoolkit/pulses/pulse_template.py]: uh... done :D

**Pull request merged**